### PR TITLE
Resizing Layer

### PR DESCRIPTION
**Pull request by koyykdy**
The branch and commit histories have been cleaned up for the PR.

IMPORT NOTE:
The lower-level op implementation of image resizing-nearest neighbor in TensorFlow.js differs from the implementation of the comparable op in Keras-Python.

While the Python version of the op function always selects the bottom right cell of the sub-matrix to be used as the representative value of that region in the downscaled matrix, the JavaScript implementation defaults to the top left cell of the sub-matrix, and then preferentially shifts to the right side of the sub-matrix in all sub-matrices past the lateral halfway point ( calculated by floor((length-1)/2) ), and the bottom side of the sub-matrix in all sub-matrices past the vertical halfway point, when considering the top-left side of the parent matrix as the origin.

This causes a slight variation in the output values from nearest neighbor downscaling between the Python and JavaScript versions of the code as it currently stands, and the unit tests for the resizing layer has been implemented to reflect this difference in op-function behavior.

Co-authored-by: Adam Lang (@AdamLang96) <adamglang96@gmail.com>
Co-authored-by: Brian Zheng (@Brianzheng123) <brianzheng345@gmail.com>

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6879)

### CODE REVIEW

**Comment by koyykdy** [comment]
@ahmedsabie @mattsoulanille Suggestions have been implemented, requesting review. Thank you!

**Comment by mattsoulanille** [comment]
This new PR has changes from packages other than tfjs-layers. Is that intentional?

We don't mind PRs having a messy commit history since we squash all commits into a single one when merging. Can we re-open your original PR #6858? This also helps me review your code since I can see changes since my last review.

If you want help with git history (e.g. getting just your changes in the PR), I can reset the `resizing` branch so it includes only files from tfjs-layers.

**Comment by koyykdy** [comment]
> The only changes aside from the `tfjs-layers/src/layers/preprocessing` should be in `tfjs-layers/src/exports-layers.ts` for registering the new layer. Any other changes would be from pulling in the new commits made by third parties to the tfjs master repository late last evening and this morning, which made the base branch for this PR outdated.

These are the relevant commits in the pull request: 
These commits are pulling in changes to the master branch to silence `This branch is out-of-date with the base branch` warning in GitHub:
https://github.com/CodeSmithDSMLProjects/tfjs/pull/41
https://github.com/CodeSmithDSMLProjects/tfjs/pull/43
Pull #43 is effectively the same as pull #41, I accidentally did it twice because doing it the first time did not make the `This branch is out-of-date with the base branch` warning go away.

This is the only commit containing the resizing layer changes:
https://github.com/CodeSmithDSMLProjects/tfjs/pull/42

**Review by mattsoulanille — APPROVED** [review]
LGTM. Thanks!

**Comment by mattsoulanille** [comment]
@ahmedsabie Please take a look at this PR when you get a chance. Thanks!

File: tfjs-layers/src/layers/preprocessing/image_resizing.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import {getExactlyOneShape} from '../../utils/types_utils';  //, getExactlyOneTensor
+
+// tf methods unimplemented in tfjs: 'bicubic', 'area', 'lanczos3', 'lanczos5',
+//                                   'gaussian', 'mitchellcubic'
+const INTERPOLATION_METHODS = new Set(['bilinear', 'nearest']);
+
+export declare interface ResizingArgs extends LayerArgs {
+  height: number;
+  width: number;
+  interpolation?: string;       // default = 'bilinear';
```
[the remark this answers is not in the document]
Comment (koyykdy) [reply tfjs-layers/src/layers/preprocessing/image_resizing.ts:26]: `// tf methods unimplemented in tfjs: 'bicubic', 'area', 'lanczos3', 'lanczos5',
//                                   'gaussian', 'mitchellcubic'
const INTERPOLATION_METHODS = new Set(['bilinear', 'nearest']);
const INTERPOLATION_KEYS = ['bilinear', 'nearest'] as const;
type INTERPOLATION_TYPE = typeof INTERPOLATION_KEYS[number];

export declare interface ResizingArgs extends LayerArgs {
  height: number;
  width: number;
  interpolation?: INTERPOLATION_TYPE; // default = 'bilinear';
  cropToAspectRatio?: boolean;        // default = false;
}`
would this be alright?

[the remark this answers is not in the document]
Comment (koyykdy) [reply tfjs-layers/src/layers/preprocessing/image_resizing.ts:26]: Got it, thank you! I think it is better not use the ternary in the type declaration for the sake of readability.

**Comment by koyykdy** [comment]
@ahmedsabie Hello, the build is failing for an unknown reason, could you check it over if you can?

**Comment by mattsoulanille** [comment]
@koyykdy Sorry about that build failure. We had changed some cloudbuild permissions on our CI runner, and it broke presubmits. It should be fixed now, though.

**Comment by koyykdy** [comment]
@mattsoulanille It seems to be failing still, could you please check what the output from the presubmit failure is? Thank you.

**Comment by mattsoulanille** [comment]
Looks like it failed to compile:
```
tfjs-layers/src/exports_layers.ts:1803:17 - error TS6133: 'categoryEncoding' is declared but its value is never read.

1803 export function categoryEncoding(args: CategoryEncodingArgs) {
                     ~~~~~~~~~~~~~~~~
```
That's strange because usually exporting a value makes typescript not try to check if it's used.

**Comment by koyykdy** [comment]
> Looks like it failed to compile:
> 
> ```
> tfjs-layers/src/exports_layers.ts:1803:17 - error TS6133: 'categoryEncoding' is declared but its value is never read.
> 
> 1803 export function categoryEncoding(args: CategoryEncodingArgs) {
>                      ~~~~~~~~~~~~~~~~
> ```
> 
> That's strange because usually exporting a value makes typescript not try to check if it's used.

This seems to be from the updates brought in by merging in from the master branch, could there be an issue with regards to this?

**Comment by koyykdy** [comment]
I attempted to update the branch, and resolved a merge conflict in exports_layers, but the rest of the changes from the main branch merge request might not have gone through.

**Comment by mattsoulanille** [comment]
Fixed it. Missing a close curly brace.

**Comment by mattsoulanille** [comment]
The build succeeds now, but the tests still have a compilation error:
```
tfjs-layers/src/layers/preprocessing/image_resizing_test.ts:113:31 - error TS2345: Argument of type '{ height: number; width: number; interpolation: string; }' is not assignable to parameter of type 'ResizingArgs'.
  Types of property 'interpolation' are incompatible.
    Type 'string' is not assignable to type 'InterpolationFormat'.

113     expect(() => new Resizing(incorrectArgs)).toThrowError(expectedError);
                                  ~~~~~~~~~~~~~

```

File: tfjs-layers/src/layers/preprocessing/image_resizing_test.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+  it('Throws an error given incorrect parameters', () => {
+    // pass incorrect interpolation method string to layer init
+    const height = 16;
+    const width = 16;
+    const interpolation = 'unimplemented';
+    const incorrectArgs = {height, width, interpolation};
+    const expectedError =
+        `Invalid interpolation parameter: ${interpolation} is not implemented`;
+    expect(() => new Resizing(incorrectArgs)).toThrowError(expectedError);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/image_resizing_test.ts:113]: Assert the type to pass invalid args in this test.

```suggestion
    expect(() => new Resizing(incorrectArgs as ResizingArgs)).toThrowError(expectedError);
```

**Comment by mattsoulanille** [comment]
@ahmedsabie Please take another look when you get a chance. Thanks!

**Review by ahmedsabie — APPROVED** [review]

**Comment by mattsoulanille** [comment]
I'm out of office at the moment, so I'll refer you to @Linchenn.

**Comment by Linchenn** [comment]
I think you are a contributor since you have merged a commit to our default branch. Probably you are not displayed in the contributor list. https://docs.github.com/en/repositories/viewing-activity-and-data-for-your-repository/viewing-a-projects-contributors @koyykdy

**Pull request merged**